Adding voice mode to Cline

arafatkatze opened the pull request:
### Testing: Please use Staging to test this in the debugger

### Related Issue

<!-- Replace XXXX with the issue number that this PR addresses -->
**Issue:** #6208 (self-referential, or create a separate issue if needed)

### Description

This PR introduces a voice dictation feature to Cline, allowing users to transcribe speech to text directly in the chat input. The implementation follows Cline's established architecture patterns using protobuf-based RPC communication between the webview and extension host.

**Key features added:**
- Speech-to-text transcription using Cline's transcription service (requires Cline account)
- Cross-platform audio recording using FFmpeg
- Real-time recording status indicators with duration tracking
- Maximum 5-minute recording limit with auto-stop
- Comprehensive error handling with user-friendly messages

**Technical approach:**
- Follows Cline's protobuf development pattern with new `DictationService` RPC endpoints
- Integrates with existing authentication and account services
- Uses FFmpeg for cross-platform audio capture (macOS, Linux, Windows)
- Implements proper state management through the existing StateManager system
- Includes telemetry tracking for usage analytics

### Test Procedure

**Manual testing performed:**
1. **Basic functionality:**
   - Tested recording start/stop/cancel operations
   - Verified transcription accuracy in English and multiple languages
   - Confirmed 5-minute auto-stop functionality works correctly

2. **Error handling:**
   - Tested without FFmpeg installed - proper error message displayed
   - Tested without Cline account authentication - appropriate auth prompt shown
   - Tested with insufficient credits - clear error message displayed
   - Tested network disconnection scenarios - connection errors handled gracefully

3. **Cross-platform testing:**
   - Verified FFmpeg detection and installation prompts on different platforms

4. **Integration testing:**
   - Confirmed settings persistence across sessions
   - Verified telemetry events are captured correctly

**Areas that need testing by reviewers:**
- Works in Both Personal Account and Org Account 
- Pricing shows up in the activity 
- Pricing shows up in the dashboard 
- You get charged nothing for cancelling a currently running run
- Installs FFMpeg cleanly


### Type of Change

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [X] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [X] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [ ] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`) - **Note: Linting issues need to be addressed**
-   [X] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [X] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

**Voice Recording UI States:**
- Idle state: Microphone icon in chat input
- Recording state: Red stop button with duration counter and cancel option
- Processing state: Loading spinner during transcription
- Error state: Error icon with tooltip showing error message

**Settings Panel:**
- New "Dictation" section in Feature Settings
- Language dropdown with 50+ supported languages
- Enable/disable toggle for the feature

*(Screenshots would be added here showing the actual UI states)*

### Additional Notes


**Dependencies:**
- Requires FFmpeg for audio recording (installation prompts provided for each platform)
- Requires active Cline account with transcription credits
- Uses Cline's transcription service API (billed at $0.006 per minute)

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Adds voice dictation feature to Cline with audio recording, transcription services, and UI integration.
> 
>   - **Behavior**:
>     - Adds voice dictation feature to Cline, enabling speech-to-text transcription in chat input.
>     - Uses FFmpeg for cross-platform audio recording with a 5-minute limit.
>     - Implements real-time recording status and error handling.
>   - **Services**:
>     - Introduces `AudioRecordingService` for handling audio capture and `VoiceTranscriptionService` for transcription.
>     - Adds `transcribeAudio()` to `ClineAccountService` for transcription API calls.
>   - **UI Components**:
>     - Updates `ChatTextArea.tsx` to include `VoiceRecorder` component for recording control.
>     - Adds `VoiceRecorder.tsx` for managing recording and transcription states.
>     - Updates `FeatureSettingsSection.tsx` to include dictation settings.
>   - **Configuration**:
>     - Adds `DictationSettings` to manage feature flags and user preferences.
>     - Updates `audioProgramConstants.ts` for FFmpeg configuration across platforms.
>   - **Telemetry**:
>     - Extends `TelemetryService` to track dictation events like recording start/stop and transcription success/error.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 9014acdb8df98523a17ed5752fdfef4f0b73d023. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

JicLotus commented on src/services/dictation/AudioRecordingService.ts:
```diff
@@ -0,0 +1,271 @@
+			}
+
+			// Create temporary file for audio output
+			const tempDir = os.tmpdir()
```
Instead of creating a temporary recording, shouldn’t we reuse our existing `data` directory for that purpose? Like gathering from `context: vscode.ExtensionContext` . Maybe we can add something around `disk.ts`

JicLotus commented on proto/cline/dictation.proto:
```diff
@@ -0,0 +1,41 @@
+  string language = 3;
+}
+
+message RecordingResult {
```
qq; Do we usually include `ErrorCode` ?

```
enum ErrorCode {
  NONE = 0;
  UNKNOWN_ERROR = 1;
....
}

message RecordingResult { 
  ErrorCode error_code = 1;
  optional string error = 2;
 }
```

That way, detailed error messages can remain optional—even when an error occurs. Also, I think we can safely drop the bool success field altogether.`.

JicLotus commented on src/core/controller/dictation/cancelRecording.ts:
```diff
@@ -0,0 +1,35 @@
+	const recordingStatus = audioRecordingService.getRecordingStatus()
+	const recordingDuration = recordingStatus.durationSeconds * 1000 // Convert to milliseconds
+
+	try {
+		const result = await audioRecordingService.cancelRecording()
+
+		telemetryService.captureVoiceRecordingStopped(taskId, recordingDuration, false, process.platform)
+
+		return RecordingResult.create({
+			success: result.success,
+			error: result.error ?? "",
+		})
+	} catch (error) {
+		console.error("Error canceling recording:", error)
+
+		telemetryService.captureVoiceRecordingStopped(taskId, recordingDuration, false, process.platform)
+
+		return RecordingResult.create({
+			success: false,
+			error: error instanceof Error ? error.message : "Unknown error occurred",
+		})
+	}
```
I would refactor to this.

```
let errorMessage = "";
let isSuccess = true;

try {
  const result = await audioRecordingService.cancelRecording();
  isSuccess = !!result?.success;
  errorMessage = result?.error ?? "";
} catch (err) {
  console.error("Error canceling recording:", err);
  isSuccess = false;
  errorMessage = err instanceof Error ? err.message : "Unknown error occurred";
}

// Track telemetry for stopping the voice recording
telemetryService.captureVoiceRecordingStopped(
  taskId,
  recordingDuration,
  false, // userCancelled
  process.platform
);

return RecordingResult.create({
  success: isSuccess,
  error: errorMessage,
});

```

JicLotus commented on src/services/dictation/AudioRecordingService.ts:
```diff
@@ -0,0 +1,271 @@
+		}
+
+		// 1. Check if the command is in the system's PATH
+		const pathDirs = (process.env.PATH || "").split(path.delimiter)
```
If you have a moment, can I get more context about this env variable `process.env.PATH` ?

arafatkatze replied:
<img width="541" height="490" alt="image" src="https://github.com/user-attachments/assets/2512c855-c1fa-4e08-a713-98d3bf33c0ae" />

Happy to explain more

JicLotus commented on src/core/controller/dictation/startRecording.ts:
```diff
@@ -0,0 +1,59 @@
+		const errorMessage = error instanceof Error ? error.message : "Unknown error occurred"
+
+		const signInAction = "Sign in to Cline"
+		const action = await HostProvider.window.showMessage({
+			type: ShowMessageType.ERROR,
+			message: `Voice recording error: ${errorMessage}`,
+			options: { items: [signInAction] },
+		})
```
Is it possible for this to throw an error as well? If so, the error will propagate unless it's caught. We should probably wrap it in a try/catch to handle that scenario safely.

JicLotus reviewed (commented):
Thanks for this PR @arafatkatze 🎉  ! I left some comments if you want to take a look.

abeatrix reviewed (commented):
Should we disable the voice mode button if the user is not signed in to Cline?

<img width="1182" height="291" alt="image" src="https://github.com/user-attachments/assets/619dac9c-9a24-4e20-a533-d07d382a653d" />

The icon is still in error state after i signed in - not auto-refreshed:

<img width="630" height="207" alt="image" src="https://github.com/user-attachments/assets/fb91e5b7-4550-43df-aefa-69a4d0d4d98a" />

It'd be great to show that dependency is missing instead of clicking on the icon that would automatically start a new task that installs the package:

<img width="704" height="955" alt="image" src="https://github.com/user-attachments/assets/5a871e2b-6be8-4d9f-a30f-3e6bf58b8682" />

The UX can use a little polish but the feature is awesome! Thanks for working on this! 

<img width="709" height="196" alt="image" src="https://github.com/user-attachments/assets/41b5b3e3-3d01-4d2f-b9ab-9f582b82b691" />

arafatkatze commented on src/services/dictation/AudioRecordingService.ts:
```diff
@@ -0,0 +1,286 @@
+			Logger.info("Stopping audio recording...")
+
+			// Send SIGINT to stop recording gracefully (like Ctrl+C)
+			this.recordingProcess.kill("SIGINT")
+
+			// Wait for the process to finish
+			await new Promise<void>((resolve) => {
+				if (this.recordingProcess) {
+					// Timeout after 5 seconds
+					const timeoutId = setTimeout(() => {
+						resolve()
+					}, 5000)
+
+					this.recordingProcess.on("exit", () => {
+						clearTimeout(timeoutId) // Clear the timeout since process exited
+						resolve()
+					})
+				} else {
+					resolve()
+				}
+			})
```
Done

arafatkatze commented on src/services/dictation/AudioRecordingService.ts:
```diff
@@ -0,0 +1,286 @@
+				}
+			})
+
+			this.recordingProcess = null
```
Done

arafatkatze commented on src/services/dictation/AudioRecordingService.ts:
```diff
@@ -0,0 +1,286 @@
+			}
+
+			// Also clean up if we have a process but isRecording is false (shouldn't happen but defensive)
+			if (!this.isRecording && this.recordingProcess) {
+				Logger.warn("Found orphaned recording process, cleaning up...")
+				this.cleanup()
+			}
```
Makes sense

arafatkatze commented on src/services/dictation/VoiceTranscriptionService.ts:
```diff
@@ -0,0 +1,146 @@
+			Logger.info("Transcription successful")
+
+			// Capture telemetry with account type - use dynamic import to avoid circular dependency
+			try {
+				const { telemetryService } = await import("@/services/telemetry")
+				telemetryService.captureVoiceTranscriptionCompleted(
+					undefined, // taskId
+					result.text?.length,
+					undefined, // duration
+					language,
+					isOrgAccount,
+				)
+			} catch (e) {
+				// Telemetry is optional, don't fail if it's not available
+				Logger.warn(`Could not capture telemetry for voice transcription: ${e}`)
+			}
```
fixed

arafatkatze commented on src/shared/ExtensionMessage.ts:
```diff
@@ -36,6 +36,8 @@ export interface ExtensionState {
 	apiConfiguration?: ApiConfiguration
 	autoApprovalSettings: AutoApprovalSettings
 	browserSettings: BrowserSettings
+	dictationSettings: DictationSettings
+	isDictationFeatureEnabled?: boolean // Feature flag for dictation
```
I assume you meant set it to true and I did that

arafatkatze commented on src/core/task/index.ts:
```diff
@@ -1960,6 +1960,11 @@ export class Task {
+		if (isFirstRequest) {
+			const durationMs = Math.round(performance.now() - this.taskInitializationStartTime)
+			telemetryService.captureTaskInitialization(this.ulid, this.taskId, durationMs, this.enableCheckpoints)
+		}
```
Rebasing horrors

arafatkatze commented on src/extension.ts:
```diff
@@ -570,6 +570,10 @@ async function getBinaryLocation(name: string): Promise<string> {
+	// Clean up audio recording service to ensure no orphaned processes
+	const { audioRecordingService } = await import("./services/dictation/AudioRecordingService")
+	audioRecordingService.cleanup()
+
```
Fixed

arafatkatze commented:
_The icon is still in error state after i signed in - not auto-refreshed:_ 

Fixed 


_It'd be great to show that dependency is missing instead of clicking on the icon that would automatically start a new task that installs the package:_

Fixed now we get

<img width="1575" height="310" alt="image" src="https://github.com/user-attachments/assets/f7f2a417-27a2-4c0a-abc3-6624ad7e54f6" />



The UX can use a little polish but the feature is awesome! Thanks for working on this!

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1672,6 +1712,55 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
 								flexDirection: "row",
 								alignItems: "center",
 							}}>
+							{dictationSettings?.dictationEnabled === true && isDictationFeatureEnabled && (
```
Why don't we set dictationSettings?.dictationEnabled as false if isDictationFeatureEnabled is false, so we can just check one value instead of two?

```suggestion
							{dictationSettings?.dictationEnabled === true && isDictationFeatureEnabled && (
```

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1672,6 +1712,55 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+									onProcessingStateChange={(isProcessing, message) => {
+										if (isProcessing && message) {
+											// Show processing message in input
+											const processingText = inputValue + (inputValue ? " " : "") + `[${message}]`
+											setInputValue(processingText)
```
nit
```suggestion
											setInputValue(`${inputValue} [${message}]`.trim())
```

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1672,6 +1712,55 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+											onSend()
+										}
+									}}
+									style={{ fontSize: 15 }}></div>
```
Seeing we are using 16.5 below, where does this 15 fontsize comes from?
```suggestion
									style={{ fontSize: 15 }}></div>
```

abeatrix commented on webview-ui/src/components/chat/VoiceRecorder.tsx:
```diff
@@ -0,0 +1,271 @@
+			</HeroTooltip>
+			<HeroTooltip content="Cancel Recording" placement="top">
+				<div
+					className={`input-icon-button text-base mt-1 text-[var(--vscode-textForeground)] ${disabled || isProcessing ? "disabled" : ""}`}
```
is vscode-textForeground valid? Alternative you can use `text-foreground` as `vscode-foreground`
```suggestion
					className={`input-icon-button text-base mt-1 text-foreground ${disabled || isProcessing ? "disabled" : ""}`}
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
+											</VSCodeOption>
+										))}
+									</VSCodeDropdown>
+									<p className="text-xs mt-1 text-[var(--vscode-descriptionForeground)]">
```
```suggestion
									<p className="text-xs mt-1 text-description">
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
+							{dictationSettings?.dictationEnabled && (
+								<div style={{ marginTop: 10, marginLeft: 20 }}>
+									<label
+										className="block text-sm font-medium text-[var(--vscode-foreground)] mb-1"
```
You can check out webview-ui/tailwind.config.mjs to see what color is available as tailwind 
```suggestion
										className="block text-sm font-medium text-foreground mb-1"
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
+									</p>
+								</div>
+							)}
+							{/* </CollapsibleContent> */}
```
```suggestion
```

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1672,6 +1712,55 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+									}}
+									style={{ fontSize: 15 }}></div>
+							)}
 							{/* <div
```
Why do we want to keep commented out code with git available

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1443,6 +1446,43 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+							speed={1}
+							spotSize={0.4}
+							spots={3}
+							style={{
+								width: "100%",
+								height: "100%",
+							}}
```
```suggestion
							className="w-full h-full"
```

abeatrix commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -1443,6 +1446,43 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
 						// Drag-over styles moved to DynamicTextArea
 						transition: "background-color 0.1s ease-in-out, border 0.1s ease-in-out",
 					}}>
+					<div
+						style={{
+							position: "absolute",
+							inset: "10px 15px", // match textarea/hightlight inset
+							pointerEvents: "none",
+							zIndex: 1, // Above textarea but below menus
+							overflow: "hidden", // clip shader to rounded rect
+							borderRadius: 2, // match textarea radius
+							transition: "opacity 1s ease-in-out",
+							opacity: isVoiceRecording ? 1 : 0,
+						}}>
```
Can't we move these to the div above? Or remove the empty div at the first level?
```suggestion
					<div
						style={{
							position: "absolute",
							inset: "10px 15px", // match textarea/hightlight inset
							pointerEvents: "none",
							zIndex: 1, // Above textarea but below menus
							overflow: "hidden", // clip shader to rounded rect
							borderRadius: 2, // match textarea radius
							transition: "opacity 1s ease-in-out",
							opacity: isVoiceRecording ? 1 : 0,
						}}>
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
+									}}>
+									Enable Dictation
+								</VSCodeCheckbox>
+								<p className="text-xs text-[var(--vscode-descriptionForeground)] mt-1">
```
```suggestion
								<p className="text-xs text-description mt-1">
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
 					)}
+					{isDictationFeatureEnabled && (
+						<>
+							<div style={{ marginTop: 10 }}>
```
```suggestion
							<div className="mt-2.5">
```

abeatrix commented on webview-ui/src/components/settings/sections/FeatureSettingsSection.tsx:
```diff
@@ -169,6 +172,63 @@ const FeatureSettingsSection = ({ renderSectionHeader }: FeatureSettingsSectionP
+							{/* TODO: Fix and use CollapsibleContent, the animation is good but it breaks the dropdown
+							<CollapsibleContent isOpen={dictationSettings?.dictationEnabled}> */}
+							{dictationSettings?.dictationEnabled && (
+								<div style={{ marginTop: 10, marginLeft: 20 }}>
```
```suggestion
								<div className="mt-2.5 ml-5">
```

abeatrix commented on webview-ui/src/components/chat/VoiceRecorder.tsx:
```diff
@@ -0,0 +1,271 @@
+			}
+
+			if (!response.audioBase64) {
+				console.error("No audio data received")
```
```suggestion
```

abeatrix commented on webview-ui/src/components/chat/VoiceRecorder.tsx:
```diff
@@ -0,0 +1,271 @@
+			)
+
+			if (transcriptionResponse.error) {
+				console.error("Transcription error:", transcriptionResponse.error)
```
If we are already setting Error then we won't need to console log again?
```suggestion
```

arafatkatze committed:
Adding vocie mode

arafatkatze committed:
Adding vocie mode

arafatkatze committed:
Update src/shared/DictationSettings.ts

Co-authored-by: ellipsis-dev[bot] <65095814+ellipsis-dev[bot]@users.noreply.github.com>

abeatrix committed:
replace styled ControlsContainer with Tailwind classes

Remove unused styled-component ControlsContainer and replace with Tailwind utility classes. Also adjust textarea positioning and padding for better visual consistency.

abeatrix committed:
Merge branch 'arafatkatze/voice-mode' of https://github.com/cline/cline into arafatkatze/voice-mode

abeatrix committed:
Fix flashing border

arafatkatze committed:
Fixing pulsing border

arafatkatze committed:
Update webview-ui/src/components/chat/ChatTextArea.tsx

Co-authored-by: ellipsis-dev[bot] <65095814+ellipsis-dev[bot]@users.noreply.github.com>

arafatkatze committed:
Fixing pulsing border

abeatrix reviewed (approved):
Verified manually:

When voice mode disabled in settings, i dont see the voice input logo:
<img width="551" height="200" alt="image" src="https://github.com/user-attachments/assets/2a9eae8a-c749-4441-9578-78fa2a337091" />

When voice mode enabled but not signed in, I'm prompted to sign in: 
<img width="1033" height="218" alt="image" src="https://github.com/user-attachments/assets/a895aadc-ce22-4763-88d9-5edcc878773d" />

Voice input works as expected:

<img width="501" height="174" alt="image" src="https://github.com/user-attachments/assets/382b893a-fb72-4af4-9e0c-c12b8770cf1a" />

gamidirohan commented:
@abeatrix Hey, where can I find this settings? I can't find it in the current Cline version.
This would make my work a lot easier. Thanks!
